### PR TITLE
Stateful tests with Dataset

### PR DESCRIPTION
**Pull request by dcherian**
I was curious to see if the hypothesis stateful test would catch an inconsistent sequence of index manipulation operations like #8646. Turns out `rename_vars` is basically broken? :P

```
E           state = DatasetStateMachine()
E           state.assert_invariants()
E           > ===
E
E            <xarray.Dataset>
E           Dimensions:  ()
E           Data variables:
E               *empty*
E           ===
E
E
E           > vars: ('1', '1_')
E           state.add_dim_coord(var=<xarray.Variable (1: 1)>
E           array([0], dtype=uint32))
E           state.assert_invariants()
E           > ===
E
E            <xarray.Dataset>
E           Dimensions:  (1: 1)
E           Coordinates:
E             * 1        (1) uint32 0
E           Data variables:
E               1_       (1) uint32 0
E           ===
E
E
E           > renaming 1 to 0
E           state.rename_vars(newname='0')
E           state.assert_invariants()
E           > ===
E
E            <xarray.Dataset>
E           Dimensions:  (1: 1)
E           Coordinates:
E             * 0        (1) uint32 0
E           Dimensions without coordinates: 1
E           Data variables:
E               1_       (1) uint32 0
E           ===
E
E
E           state.teardown()
```

### CODE REVIEW

**Comment by max-sixty** [comment]
Great idea! Really cool.

I even wonder to what extent this could replace vast swathes of our tests — anything where the output is calculated from the input.

We'll always want some unit tests to code to when developing, but plausibly the unit tests are either a) examples or b) some saved cases that have failed these model-based tests before.

**Comment by dcherian** [comment]
>  instead some valid cases require skipping the default indexes invariant check.

Can we write down rules for when these checks are needed?

**Comment by benbovy** [comment]
Yes although I think it is easier to write rules for when these checks aren't needed.

One example: skip the default indexes invariant test when the name of an existing dimension coordinate is passed as input kwarg or dict key to `.rename_vars()`.

**Comment by dcherian** [comment]
This seems to have found at least 3 edge-case failures for `swap_dims`:
1. swapping a dimension to itself
2. swapping from Index to a MultiIndex level
3. swapping from MultiIndex to a level of the same MultiIndex

File: properties/test_index_manipulation.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+
+@st.composite
+def unique(draw, strategy):
+    # https://stackoverflow.com/questions/73737073/create-hypothesis-strategy-that-returns-unique-values
+    seen = draw(st.shared(st.builds(set), key="key-for-unique-elems"))
+    return draw(
+        strategy.filter(lambda x: x not in seen).map(lambda x: seen.add(x) or x)
+    )
+
```
Comment (TomNicholas) [inline properties/test_index_manipulation.py:50]: Presumably you could just use the `xrst.unique_subset_of` strategy with `minsize=1` and `maxsize=1` instead here?

File: xarray/testing/strategies.py [hunk 2]
```diff
@@ -192,6 +197,22 @@ def attrs() -> st.SearchStrategy[Mapping[Hashable, Any]]:
     )
 
 
+def serializable_attrs() -> st.SearchStrategy[Mapping[Hashable, Any]]:
```
Comment (TomNicholas) [inline xarray/testing/strategies.py:200]: What makes the current attrs strategy produce non-serializable things? TBH that `attrs` strategy is the least important part of the existing `strategies` module, and generating nested things is probably taking time up for little coverage benefit, so I wouldn't be opposed to simplifying the main `attrs` strategy so you can use it here.

File: properties/test_index_manipulation.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+random.seed(123456)
+
+# Share to ensure we get unique names on each draw?
+UNIQUE_NAME = unique(strategy=xrst.names())
+DIM_NAME = xrst.dimension_names(name_strategy=UNIQUE_NAME, min_dims=1, max_dims=1)
```
Comment (TomNicholas) [inline properties/test_index_manipulation.py:56]: What do you mean by "share" here?

**Review by TomNicholas — COMMENTED** [review]
I'm glad the strategies we added are being built upon!

File: properties/test_index_manipulation.py (hunk 3 above)
Comment (dcherian) [reply properties/test_index_manipulation.py:56]: I needed to draw a value from `names`   that is not present in the dataset for stack. 

Alternatively, I could use `assume(newname not in self.dataset)`

File: properties/test_index_manipulation.py (hunk 1 above)
Comment (dcherian) [reply properties/test_index_manipulation.py:50]: ah `unique_subset_of` is indeed useful, but not for this case I don't think. Repeated draws can result in duplicated values, and I wanted to avoid that.

File: properties/test_index_manipulation.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    @precondition(lambda self: len(self.dataset._variables) >= 2)
+    def swap_dims(self, data, dim):
+        ds = self.dataset
+        choices = [name for name, var in ds._variables.items() if var.dims == (dim,)]
+        # TODO: is there a better way to skip if choices == []
+        # note(choices)
+        # if not choices:
+        #     return dim
+        # Can only swap to a variable with the same dim
+        to = data.draw(st.sampled_from(choices))
```
Comment (dcherian) [inline properties/test_index_manipulation.py:133]: @Zac-HD quick q: I need to skip tests when choices is empty but using	`filter` or `assume` tells me that this is filtering too much. Is there some other approach, or shall I just disable the filter_too_much check?

File: properties/test_index_manipulation.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    mindexes = [
+        name
+        for name, index in ds.xindexes.items()
+        if isinstance(index, PandasMultiIndex)
+    ]
+    return mindexes
+
+
+def get_dimension_coordinates(ds: Dataset) -> tuple[Hashable]:
+    return tuple(set(ds.dims) & set(ds._variables))
```
Comment (Zac-HD) [inline properties/test_index_manipulation.py:47]: Unused code?  Also note that going via `set()` can make tests flaky, via nondeterministic iteration order.

File: properties/test_index_manipulation.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+@st.composite
+def unique(draw, strategy):
+    # https://stackoverflow.com/questions/73737073/create-hypothesis-strategy-that-returns-unique-values
+    seen = draw(st.shared(st.builds(set), key="key-for-unique-elems"))
+    return draw(
+        strategy.filter(lambda x: x not in seen).map(lambda x: seen.add(x) or x)
+    )
+
+
+random.seed(123456)
```
Comment (Zac-HD) [inline properties/test_index_manipulation.py:59]: What's this intended to do?

File: properties/test_index_manipulation.py (hunk 4 above)
Comment (Zac-HD) [reply properties/test_index_manipulation.py:133]: I think ignoring it would lead to very slow tests, so let's try to make this fast.

The obvious way to do _that_ is to have a precondition that there exist (dimensions with) valid choices among the variables, and then to arrange your `dims` strategy to choose such a dimension.  

Unfortunately there's currently no way to put a filter on a `consumes(bundle)` strategy, which seems something of an oversight - it'd just take a few dozen lines of code.  Until someone does that, I'd probably just turn your `indexed_dims` bundle into an instance variable - it's also usable in a precondition that way, and you can `draw(st.sampled_from(self.indexed_dims))` inside the test easily enough.

File: xarray/testing/strategies.py [hunk 7]
```diff
@@ -62,6 +64,15 @@ def supported_dtypes() -> st.SearchStrategy[np.dtype]:
     )
 
 
+# TODO: add datetime64[ns]
+def pandas_index_dtypes() -> st.SearchStrategy[np.dtype]:
+    return (
+        npst.integer_dtypes(endianness="<", sizes=(32, 64))
+        | npst.unsigned_integer_dtypes(endianness="<", sizes=(32, 64))
+        | npst.floating_dtypes(endianness="<", sizes=(32, 64))
```
Comment (Zac-HD) [inline xarray/testing/strategies.py:72]: Really little-endian, or 'machine-native byteorder happens to be' little-endian?  A comment to distinguish would be nice.

File: xarray/testing/strategies.py [hunk 8]
```diff
@@ -361,6 +393,23 @@ def variables(
...
+        st.SearchStrategy[Union[Sequence[Hashable], Mapping[Hashable, int]]],
+        None,
+    ] = None,
+    dtype: st.SearchStrategy[np.dtype] = pandas_index_dtypes(),
+    attrs: st.SearchStrategy[Mapping] = attrs(),
+) -> xr.Variable:
+
+    index = draw(pdst.indexes(min_size=1, dtype=draw(dtype)))
+    _dims = draw(dimension_names(min_dims=1, max_dims=1))
+    return xr.Variable(dims=_dims, data=index, attrs=draw(attrs))
```
Comment (Zac-HD) [inline xarray/testing/strategies.py:410]: This could be a function which returns a `st.builds(xr.Variable, ...)` rather than a composite, which is a _slight_ performance improvement but mostly just an idiom thing.

**Comment by dcherian** [comment]
Hah now our strategy tests are failing

File: properties/test_index_manipulation.py [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+    # dim coord
+    ds[name] = var
+    # non-dim coord of same size; this allows renaming
+    ds[name + "_"] = var
+
+
+class DatasetStateMachine(RuleBasedStateMachine):
+    # Can't use bundles because we'd need pre-conditions on consumes(bundle)
+    # indexed_dims = Bundle("indexed_dims")
+    # multi_indexed_dims = Bundle("multi_indexed_dims")
```
Comment (Zac-HD) [inline properties/test_index_manipulation.py:65]: I think https://github.com/HypothesisWorks/hypothesis/issues/3944 will enable this, though I wouldn't delay this PR to wait.

File: properties/test_index_manipulation.py [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+DatasetStateMachine.TestCase.settings = settings(max_examples=300, deadline=None)
+DatasetTest = DatasetStateMachine.TestCase
+
+
+@pytest.mark.skip(reason="failure detected by hypothesis")
+def test_unstack_object():
+    import xarray as xr
+
+    ds = xr.Dataset()
+    ds["0"] = np.array(["", "\x000"], dtype=object)
```
Comment (Zac-HD) [inline properties/test_index_manipulation.py:263]: It doesn't seem that surprising that you hit bugs related to empty-string or null-prefixed string?  To most C code, those are identical!

File: xarray/testing/strategies.py [hunk 11]
```diff
@@ -97,14 +119,18 @@ def dimension_names(
...
     ----------
+    name_strategy
+        Strategy for making names. Useful if we need to share this.
     min_dims
         Minimum number of dimensions in generated list.
     max_dims
         Maximum number of dimensions in generated list.
     """
 
+    elements = names() if name_strategy is None else name_strategy
```
Comment (Zac-HD) [inline xarray/testing/strategies.py:130]: You can just set `name_strategy=names()` as a default argument 🙂

**Review by Zac-HD — APPROVED** [review]
Obviously there are some remaining todo comments, but this looks great!  I'd consider merging it more or less as-is, and then continuing work in a follow-up PR 🙂

**Comment by dcherian** [comment]
Thanks for the review @Zac-HD
